src/scene/gsplat/gsplat-data.js

```diff
@@ -408,6 +412,83 @@ class GSplatData {
             })
         }], false);
     }
+
+    calcMortonOrder() {
+        const calcMinMax = (arr) => {
+            let min = arr[0];
+            let max = arr[0];
+            for (let i = 1; i < arr.length; i++) {
+                if (arr[i] < min) min = arr[i];
+                if (arr[i] > max) max = arr[i];
+            }
+            return { min, max };
+        };
+
+        // https://fgiesen.wordpress.com/2009/12/13/decoding-morton-codes/
+        const encodeMorton3 = (x, y, z) => {
+            const Part1By2 = (x) => {
+                x &= 0x000003ff;
+                x = (x ^ (x << 16)) & 0xff0000ff;
+                x = (x ^ (x <<  8)) & 0x0300f00f;
+                x = (x ^ (x <<  4)) & 0x030c30c3;
+                x = (x ^ (x <<  2)) & 0x09249249;
+                return x;
+            };
+
+            return (Part1By2(z) << 2) + (Part1By2(y) << 1) + Part1By2(x);
+        };
+
+        const x = this.getProp('x');
+        const y = this.getProp('y');
+        const z = this.getProp('z');
+
+        const { min: minX, max: maxX } = calcMinMax(x);
+        const { min: minY, max: maxY } = calcMinMax(y);
+        const { min: minZ, max: maxZ } = calcMinMax(z);
+
+        const sizeX = 1024 / (maxX - minX);
+        const sizeY = 1024 / (maxY - minY);
+        const sizeZ = 1024 / (maxZ - minZ);
+
+        const morton = new Uint32Array(this.numSplats);
+        for (let i = 0; i < this.numSplats; i++) {
+            const ix = Math.floor((x[i] - minX) * sizeX);
+            const iy = Math.floor((y[i] - minY) * sizeY);
+            const iz = Math.floor((z[i] - minZ) * sizeZ);
+            morton[i] = encodeMorton3(ix, iy, iz);
+        }
+
+        // generate indices
+        const indices = new Uint32Array(this.numSplats);
+        for (let i = 0; i < this.numSplats; i++) {
+            indices[i] = i;
+        }
+        // order splats by morton code
+        indices.sort((a, b) => morton[a] - morton[b]);
```

Comment:
If this is the expensive part, a bucket sort could be used instead.

Reply:
this isn't a fast bit, but generating the morton order is far slower. i will possibly investigate reimplementing this whole function in wasm and then also try bucketing.